File: client-frontend/src/services/postService.ts
import api from '@/lib/api';
import { mediaService, MediaResponse } from './mediaService';
import {
  Post,
  Comment,
  CreatePostRequest,
  UpdatePostRequest,
  CreateCommentRequest,
  CreateInteractionRequest,
  Interaction,
  PaginatedResponse,
  InteractionType,
  ReactionType
} from '@/types';

export const postService = {
  // Create post with proper structure matching backend
  async createPost(postData: CreatePostRequest, files?: File[]): Promise<Post> {
    try {
      let imageUrls: string[] = [];
      let videoUrls: string[] = [];

      // Step 1: Upload files to media service first if files exist
      if (files && files.length > 0) {
        console.log('Uploading files to media service...');
        const mediaResponses: MediaResponse[] = await mediaService.uploadFiles(
          files,
          'Post media files'
        );

        // Separate images and videos based on media type
        mediaResponses.forEach(media => {
          if (media.mediaType === 'IMAGE') {
            imageUrls.push(media.cloudinaryUrl);
          } else if (media.mediaType === 'VIDEO') {
            videoUrls.push(media.cloudinaryUrl);
          }
        });

        console.log('Files uploaded successfully - Images:', imageUrls, 'Videos:', videoUrls);
      }

      // Step 2: Create post data with separated media URLs
      const postRequestData = {
        ...postData,
        images: imageUrls.length > 0 ? imageUrls : undefined,
        videos: videoUrls.length > 0 ? videoUrls : undefined
      };
      console.log('Creating post request...', postRequestData);

      // Step 3: Create post via post service
      const response = await api.post('/posts', postRequestData, {
        headers: {
          'Content-Type': 'application/json',
        },
      });

      return response.data;
    } catch (error) {
      console.error('Error creating post:', error);
      throw error;
    }
  },

  // Get paginated posts
  async getPosts(
    page = 0,
    size = 10,
    sortBy = 'createdAt',
    sortDir = 'desc',
    authorId?: string,
    category?: string,
    search?: string
  ): Promise<PaginatedResponse<Post>> {
    const params = new URLSearchParams({
      page: page.toString(),
      size: size.toString(),
      sortBy,
      sortDir,
    });

    if (authorId) params.append('authorId', authorId);
    if (category) params.append('category', category);
    if (search) params.append('search', search);

    const response = await api.get(`/posts?${params.toString()}`);

    return response.data;
  },

  // Get single post by ID
  async getPost(postId: string): Promise<Post> {
    const response = await api.get(`/posts/${postId}`);
    return response.data;
  },

  // Get posts by author
  async getUserPosts(authorId: string, page = 0, size = 10): Promise<PaginatedResponse<Post>> {
    const response = await api.get(`/posts/user/${authorId}`, {
      params: {
        page,
        size
      }
    });
    console.log("Fetched user posts:", response.data);
    return response.data;
  },

  // Update post
  async updatePost(postId: string, updateData: UpdatePostRequest): Promise<Post> {
    const response = await api.put(`/posts/${postId}`, updateData);
    return response.data;
  },

  // Delete post
  async deletePost(postId: string): Promise<void> {
    await api.delete(`/posts/${postId}`);
  },

  // Search posts
  async searchPosts(query: string, page = 0, size = 10): Promise<PaginatedResponse<Post>> {
    return this.getPosts(page, size, 'createdAt', 'desc', undefined, undefined, query);
  },

  // Get posts by category
  async getPostsByCategory(category: string, page = 0, size = 10): Promise<PaginatedResponse<Post>> {
    return this.getPosts(page, size, 'createdAt', 'desc', undefined, category);
  },

  // INTERACTION METHODS

  // Create interaction (like, share, bookmark)
  async createInteraction(postId: string, interactionData: CreateInteractionRequest): Promise<Interaction | null> {
    const response = await api.post(`/posts/${postId}/interactions`, interactionData);
    console.log("Created interaction:", response.data);
    return response.data;
  },

  // Like/Unlike post
  async toggleLike(postId: string): Promise<Interaction | null> {
    return this.createInteraction(postId, {
      type: InteractionType.LIKE,
      reactionType: ReactionType.LIKE
    });
  },

  // Share post
  async sharePost(postId: string): Promise<Interaction | null> {
    return this.createInteraction(postId, {
      type: InteractionType.SHARE
    });
  },

  // Bookmark post
  async toggleBookmark(postId: string): Promise<Interaction | null> {
    return this.createInteraction(postId, {
      type: InteractionType.BOOKMARK,
      reactionType: ReactionType.BOOKMARK
    });
  },

  // Check if user has liked post
  async hasUserLikedPost(postId: string): Promise<boolean> {
    const response = await api.get(`/posts/${postId}/likes/check`);
    return response.data;
  },

  // Get top viewed posts
  async getTopViewedPosts(): Promise<Post[]> {
    const topViewedResponse = await api.get('/posts/top-viewed');
    return topViewedResponse.data;
  },

  // Get top liked posts
  async getTopLikedPosts(): Promise<Post[]> {
    const response = await api.get('/posts/top-liked');
    return response.data;
  },

  // INTERACTION STATUS METHODS (New - for persistent state)

  // Check interaction status for a post
  async getInteractionStatus(postId: string): Promise<{
    postId: string;
    userId: string;
    hasLiked: boolean;
    hasBookmarked: boolean;
    interactions: { LIKE: boolean; BOOKMARK: boolean };
  }> {
    try {
      const response = await api.get(`/posts/${postId}/interactions/status`);
      const data = response.data;

      // Add debugging to understand the actual response structure
      console.log('Backend interaction status response:', data);

      // Handle the actual backend response structure
      // Backend returns: { id, postId, userId, type, reactionType, hasInteraction, active, ... }

      let hasLiked = false;
      let hasBookmarked = false;

      // If data is an array of interactions
      if (Array.isArray(data)) {
        console.log('Processing array of interactions:', data.length);
        hasLiked = data.some(interaction => {
          const isLike = interaction.type === 'LIKE' && interaction.hasInteraction && interaction.active;
          console.log('Checking interaction for LIKE:', interaction, 'Result:', isLike);
          return isLike;
        });
        hasBookmarked = data.some(interaction => {
          const isBookmark = interaction.type === 'BOOKMARK' && interaction.hasInteraction && interaction.active;
          console.log('Checking interaction for BOOKMARK:', interaction, 'Result:', isBookmark);
          return isBookmark;
        });
      }
      // If data is a single interaction object
      else if (data && typeof data === 'object') {
        console.log('Processing single interaction object:', data);
        // Check if it's the structure you showed: { hasInteraction: true, type: "LIKE", ... }
        if (data.hasInteraction && data.active) {
          hasLiked = data.type === 'LIKE';
          hasBookmarked = data.type === 'BOOKMARK';
          console.log('Single interaction - hasLiked:', hasLiked, 'hasBookmarked:', hasBookmarked);
        }
        // Or if it's already in the expected format
        else if ('hasLiked' in data) {
          hasLiked = data.hasLiked;
          hasBookmarked = data.hasBookmarked;
          console.log('Expected format - hasLiked:', hasLiked, 'hasBookmarked:', hasBookmarked);
        }
      }

      console.log('Final interaction status:', { hasLiked, hasBookmarked });

      return {
        postId,
        userId: data?.userId || '',
        hasLiked,
        hasBookmarked,
        interactions: {
          LIKE: hasLiked,
          BOOKMARK: hasBookmarked
        }
      };
    } catch (error) {
      console.error('Error fetching interaction status:', error);
      // Return default values on error
      return {
        postId,
        userId: '',
        hasLiked: false,
        hasBookmarked: false,
        interactions: {
          LIKE: false,
          BOOKMARK: false
        }
      };
    }
  },

  // Check like status specifically
  async checkLikeStatus(postId: string): Promise<{
    postId: string;
    hasLiked: boolean;
    userId: string;
  }> {
    const response = await api.get(`/posts/${postId}/interactions/like/status`);
    return response.data;
  },

  // Check bookmark status specifically
  async checkBookmarkStatus(postId: string): Promise<{
    postId: string;
    hasBookmarked: boolean;
    userId: string;
  }> {
    const response = await api.get(`/posts/${postId}/interactions/bookmark/status`);
    return response.data;
  },

  // COMMENT METHODS (Missing)

  // Get comments for a post
  async getComments(postId: string, page = 0, size = 10): Promise<PaginatedResponse<Comment>> {
    const params = new URLSearchParams({
      page: page.toString(),
      size: size.toString(),
    });

    const response = await api.get(`/posts/${postId}/comments?${params.toString()}`);
    console.log("Fetched comments for post:", postId, response.data);
    return response.data;
  },

  // Create a comment
  async createComment(postId: string, commentData: CreateCommentRequest): Promise<Comment> {
    const response = await api.post(`/posts/${postId}/comments`, commentData);
    return response.data;
  },

  async getMyPosts(page = 0, size = 10): Promise<PaginatedResponse<Post>> {
    const response = await api.get(`/posts/me?page=${page}&size=${size}`);

    return response.data;
  },

};

File: post-service/src/main/java/com/ctuconnect/service/PostService.java
package com.ctuconnect.service;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.web.multipart.MultipartFile;
import com.ctuconnect.client.MediaServiceClient;
import com.ctuconnect.client.UserServiceClient;
import com.ctuconnect.dto.AuthorInfo;
import com.ctuconnect.dto.request.PostRequest;
import com.ctuconnect.dto.request.ScheduledPostRequest;
import com.ctuconnect.dto.response.PostResponse;
import com.ctuconnect.dto.response.PostAnalyticsResponse;
import com.ctuconnect.entity.InteractionEntity;
import com.ctuconnect.entity.PostEntity;
import com.ctuconnect.entity.CommentEntity;
import com.ctuconnect.repository.CommentRepository;
import com.ctuconnect.repository.InteractionRepository;
import com.ctuconnect.repository.PostRepository;
import com.ctuconnect.security.AuthenticatedUser;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Optional;
import java.util.Set;
import java.util.stream.Collectors;

@Service
public class PostService {

    @Autowired
    private PostRepository postRepository;

    @Autowired
    private CommentRepository commentRepository;

    @Autowired
    private InteractionRepository interactionRepository;

    @Autowired
    private MediaServiceClient mediaServiceClient;

    @Autowired
    private UserServiceClient userServiceClient;

    @Autowired
    private EventService eventService;

    public PostResponse createPost(PostRequest request, List<MultipartFile> files, String authorId) {
        AuthorInfo author = userServiceClient.getAuthorInfo(authorId);
        if (author == null) {
            throw new RuntimeException("Author not found with id: " + authorId);
        }

        PostEntity post = PostEntity.builder()
                .title(request.getTitle())
                .content(request.getContent())
                .author(author)
                .images(new ArrayList<>())
                .tags(request.getTags() != null ? request.getTags() : new ArrayList<>())
                .category(request.getCategory())
                .privacy(request.getVisibility() != null ? request.getVisibility() : "PUBLIC")
                .stats(new PostEntity.PostStats())
                .build();

        // Upload files to media-service
        if (files != null && !files.isEmpty()) {
            List<String> imageUrls = new ArrayList<>();
            for (MultipartFile file : files) {
                if (!file.isEmpty()) {
                    try {
                        MediaServiceClient.MediaUploadResponse uploadResponse =
                            mediaServiceClient.uploadFile(file, getFileType(file));
                        imageUrls.add(uploadResponse.getFileUrl());
                    } catch (Exception e) {
                        // Log error but continue processing
                        System.err.println("Failed to upload file: " + e.getMessage());
                    }
                }
            }
            post.setImages(imageUrls);
        }

        PostEntity savedPost = postRepository.save(post);

        // Publish event
        eventService.publishPostEvent("POST_CREATED", savedPost.getId(), savedPost.getAuthorId(), savedPost);

        return new PostResponse(savedPost);
    }

    public Page<PostResponse> getAllPosts(Pageable pageable) {
        Page<PostEntity> posts = postRepository.findAll(pageable);
        
        // Recalculate stats for each post before returning
        posts.forEach(this::recalculatePostStats);
        postRepository.saveAll(posts.getContent());
        
        return posts.map(PostResponse::new);
    }

    public Page<PostResponse> getPostsByAuthor(String authorId, Pageable pageable) {
        System.out.println("DEBUG: PostService.getPostsByAuthor called with authorId: " + authorId);

        Page<PostEntity> posts = postRepository.findByAuthor_Id(authorId, pageable);
        System.out.println("DEBUG: Repository query returned " + posts.getTotalElements() + " posts");

        // Debug first few posts from repository
        posts.getContent().stream().limit(3).forEach(post -> {
            System.out.println("DEBUG: Repository returned post ID: " + post.getId() +
                ", Author ID: " + (post.getAuthor() != null ? post.getAuthor().getId() : "null") +
                ", Author Name: " + (post.getAuthor() != null ? post.getAuthor().getName() : "null"));
        });

        // Recalculate stats for each post before returning
        posts.forEach(this::recalculatePostStats);
        postRepository.saveAll(posts.getContent());
        
        return posts.map(PostResponse::new);
    }

    public Page<PostResponse> getPostsByCategory(String category, Pageable pageable) {
        return postRepository.findByCategory(category, pageable)
                .map(PostResponse::new);
    }

    public Page<PostResponse> searchPosts(String searchTerm, Pageable pageable) {
        return postRepository.findByTitleContainingOrContentContaining(searchTerm, searchTerm, pageable)
                .map(PostResponse::new);
    }

    public PostResponse getPostById(String id, String currentUserId) {
        Optional<PostEntity> postOpt = postRepository.findById(id);
        if (postOpt.isPresent()) {
            PostEntity post = postOpt.get();

            // Record view interaction if user is different from author
            if (currentUserId != null && !currentUserId.equals(post.getAuthorId())) {
                recordViewInteraction(post.getId(), currentUserId);
            }

            // Recalculate stats from database before returning
            recalculatePostStats(post);
            postRepository.save(post);
            
            return new PostResponse(post);
        }
        throw new RuntimeException("Post not found with id: " + id);
    }

    /**
     * Recalculate post stats from actual interactions in database
     * This fixes the issue where stats show 0 even when interactions exist
     */
    private void recalculatePostStats(PostEntity post) {
        // Count actual likes from interactions
        long likeCount = interactionRepository.countByPostIdAndType(post.getId(), InteractionEntity.InteractionType.LIKE);
        long bookmarkCount = interactionRepository.countByPostIdAndType(post.getId(), InteractionEntity.InteractionType.BOOKMARK);
        long shareCount = interactionRepository.countByPostIdAndType(post.getId(), InteractionEntity.InteractionType.SHARE);
        
        // Count comments
        long commentCount = commentRepository.countByPostId(post.getId());
        
        // Update post stats
        post.getStats().setLikes(likeCount);
        post.getStats().setComments(commentCount); 
        post.getStats().setShares(shareCount);
        
        // Update reactions map for LIKE type
        post.getStats().getReactions().put(InteractionEntity.ReactionType.LIKE, (int) likeCount);
    }

    public PostResponse updatePost(String id, PostRequest request, String authorId) {
        Optional<PostEntity> postOpt = postRepository.findById(id);
        if (postOpt.isPresent()) {
            PostEntity post = postOpt.get();

            // Check if user is the author
            if (!post.getAuthorId().equals(authorId)) {
                throw new RuntimeException("Only the author can update this post");
            }

            if (request.getTitle() != null) {
                post.setTitle(request.getTitle());
            }
            if (request.getContent() != null) {
                post.setContent(request.getContent());
            }
            if (request.getTags() != null) {
                post.setTags(request.getTags());
            }
            if (request.getCategory() != null) {
                post.setCategory(request.getCategory());
            }
            if (request.getVisibility() != null) {
                post.setVisibility(request.getVisibility());
            }

            PostEntity savedPost = postRepository.save(post);

            // Publish event
            eventService.publishPostEvent("POST_UPDATED", savedPost.getId(), savedPost.getAuthorId(), savedPost);

            return new PostResponse(savedPost);
        }
        throw new RuntimeException("Post not found with id: " + id);
    }

    public void deletePost(String id, String authorId) {
        Optional<PostEntity> postOpt = postRepository.findById(id);
        if (postOpt.isPresent()) {
            PostEntity post = postOpt.get();

            // Check if user is the author
            if (!post.getAuthorId().equals(authorId)) {
                throw new RuntimeException("Only the author can delete this post");
            }

            // Delete associated comments and interactions
            commentRepository.deleteByPostId(id);
            interactionRepository.deleteByPostId(id);

            // Delete the post
            postRepository.deleteById(id);

            // Publish event
            eventService.publishPostEvent("POST_DELETED", id, authorId, post);
        } else {
            throw new RuntimeException("Post not found with id: " + id);
        }
    }

    private void recordViewInteraction(String postId, String userId) {
        try {
            AuthorInfo author = userServiceClient.getAuthorInfo(userId);
            if (author == null) return;

            // Check if user already viewed this post recently (within last hour)
            Optional<InteractionEntity> existingView = interactionRepository
                    .findByPostIdAndAuthor_IdAndType(postId, userId, InteractionEntity.InteractionType.VIEW);

            if (existingView.isEmpty()) {
                // Create view interaction
                InteractionEntity viewInteraction = new InteractionEntity(postId, author, InteractionEntity.InteractionType.VIEW);
                interactionRepository.save(viewInteraction);

                // Update post stats
                Optional<PostEntity> postOpt = postRepository.findById(postId);
                if (postOpt.isPresent()) {
                    PostEntity post = postOpt.get();
                    post.getStats().incrementViews();
                    postRepository.save(post);
                }

                // Publish interaction event
                eventService.publishInteractionEvent(postId, userId, "VIEW");
            }
        } catch (Exception e) {
            // Log error but don't fail the main operation
            System.err.println("Failed to record view interaction: " + e.getMessage());
        }
    }

    private String getFileType(MultipartFile file) {
        String contentType = file.getContentType();
        if (contentType != null) {
            if (contentType.startsWith("image/")) return "IMAGE";
            if (contentType.startsWith("video/")) return "VIDEO";
            if (contentType.startsWith("audio/")) return "AUDIO";
            if (contentType.equals("application/pdf")) return "PDF";
        }
        return "DOCUMENT";
    }

    public List<PostResponse> getTopViewedPosts() {
        return postRepository.findTop10ByOrderByStatsViewsDesc()
                .stream()
                .map(PostResponse::new)
                .toList();
    }

    public List<PostResponse> getTopLikedPosts() {
        return postRepository.findTop10ByOrderByStatsLikesDesc()
                .stream()
                .map(PostResponse::new)
                .toList();
    }

    /**
     * Enhanced post creation with Facebook-like features
     */
    public PostResponse createEnhancedPost(PostRequest request, AuthenticatedUser user) {
        AuthorInfo author = userServiceClient.getAuthorInfo(user.getId());
        if (author == null) {
            throw new RuntimeException("Author not found with id: " + user.getId());
        }

        PostEntity post = PostEntity.builder()
                .title(request.getTitle())
                .content(request.getContent())
                .author(author)
                .images(request.getImages() != null ? request.getImages() : new ArrayList<>())
                .videos(request.getVideos() != null ? request.getVideos() : new ArrayList<>())
                .tags(request.getTags() != null ? request.getTags() : new ArrayList<>())
                .category(request.getCategory())
                .privacy(request.getVisibility() != null ? request.getVisibility() : "PUBLIC")
                .postType(request.getPostType() != null ? PostEntity.PostType.valueOf(request.getPostType()) : PostEntity.PostType.TEXT)
                .stats(new PostEntity.PostStats())
                .audienceSettings(new PostEntity.AudienceSettings())
                .engagement(new PostEntity.EngagementMetrics())
                .build();

        // Set audience settings if provided
        if (request.getAudienceSettings() != null) {
            post.setAudienceSettings(request.getAudienceSettings());
        }

        // Handle scheduled posts
        if (request.getScheduledAt() != null) {
            post.setScheduledAt(request.getScheduledAt());
            post.setScheduled(true);
        }

        PostEntity savedPost = postRepository.save(post);

        // Publish event
        eventService.publishPostEvent("POST_CREATED", savedPost.getId(), savedPost.getAuthorId(), savedPost);

        return new PostResponse(savedPost);
    }

    /**
     * Get users affected by a post (for cache invalidation)
     */
    public Set<String> getAffectedUserIds(String postId) {
        Optional<PostEntity> postOpt = postRepository.findById(postId);
        if (postOpt.isPresent()) {
            PostEntity post = postOpt.get();
            Set<String> affectedUsers = new HashSet<>();

            // Add author
            affectedUsers.add(post.getAuthorId());

            // Add friends if post is visible to friends
            if ("FRIENDS".equals(post.getPrivacy()) || "PUBLIC".equals(post.getPrivacy())) {
                Set<String> authorFriends = userServiceClient.getFriendIds(post.getAuthorId());
                affectedUsers.addAll(authorFriends);
            }

            return affectedUsers;
        }
        return new HashSet<>();
    }

    /**
     * Handle post interactions (like, comment, share)
     */
    public void handlePostInteraction(String postId, String userId, String action, String reactionType) {
        Optional<PostEntity> postOpt = postRepository.findById(postId);
        if (postOpt.isEmpty()) {
            throw new RuntimeException("Post not found with id: " + postId);
        }

        PostEntity post = postOpt.get();
        AuthorInfo user = userServiceClient.getAuthorInfo(userId);
        if (user == null) {
            throw new RuntimeException("User not found with id: " + userId);
        }

        switch (action.toUpperCase()) {
            case "LIKE":
                handleLikeInteraction(postId, userId, reactionType, post, user);
                break;
            case "UNLIKE":
                handleUnlikeInteraction(postId, userId, post);
                break;
            case "SHARE":
                handleShareInteraction(postId, userId, post, user);
                break;
            default:
                throw new RuntimeException("Unsupported action: " + action);
        }

        // Update engagement metrics
        updateEngagementMetrics(post);
        postRepository.save(post);
    }

    /**
     * Get post author ID
     */
    public String getPostAuthorId(String postId) {
        Optional<PostEntity> postOpt = postRepository.findById(postId);
        if (postOpt.isPresent()) {
            return postOpt.get().getAuthorId();
        }
        throw new RuntimeException("Post not found with id: " + postId);
    }

    /**
     * Add comment to post
     */
    public void addComment(String postId, String userId, String content) {
        Optional<PostEntity> postOpt = postRepository.findById(postId);
        if (postOpt.isEmpty()) {
            throw new RuntimeException("Post not found with id: " + postId);
        }

        AuthorInfo author = userServiceClient.getAuthorInfo(userId);
        if (author == null) {
            throw new RuntimeException("User not found with id: " + userId);
        }

        CommentEntity comment = CommentEntity.builder()
                .postId(postId)
                .author(author)
                .content(content)
                .build();

        commentRepository.save(comment);

        // Update post comment count
        PostEntity post = postOpt.get();
        post.getStats().incrementComments();
        updateEngagementMetrics(post);
        postRepository.save(post);

        // Publish event
        eventService.publishPostEvent("COMMENT_ADDED", postId, userId, comment);
    }

    /**
     * Get post analytics
     */
    public PostAnalyticsResponse getPostAnalytics(String postId, String userId) {
        Optional<PostEntity> postOpt = postRepository.findById(postId);
        if (postOpt.isEmpty()) {
            throw new RuntimeException("Post not found with id: " + postId);
        }

        PostEntity post = postOpt.get();

        // Check if user is the post author
        if (!post.getAuthorId().equals(userId)) {
            throw new RuntimeException("Only post author can view analytics");
        }

        return PostAnalyticsResponse.builder()
                .postId(postId)
                .views(post.getStats().getViews())
                .likes(post.getStats().getLikes())
                .comments(post.getStats().getComments())
                .shares(post.getStats().getShares())
                .engagementRate(post.getEngagement().getEngagementRate())
                .reactions(post.getStats().getReactions())
                .build();
    }

    /**
     * Schedule post for later publishing
     */
    public PostResponse schedulePost(ScheduledPostRequest request, AuthenticatedUser user) {
        AuthorInfo author = userServiceClient.getAuthorInfo(user.getId());
        if (author == null) {
            throw new RuntimeException("Author not found with id: " + user.getId());
        }

        PostEntity post = PostEntity.builder()
                .title(request.getTitle())
                .content(request.getContent())
                .author(author)
                .images(request.getImages() != null ? request.getImages() : new ArrayList<>())
                .tags(request.getTags() != null ? request.getTags() : new ArrayList<>())
                .category(request.getCategory())
                .privacy(request.getVisibility() != null ? request.getVisibility() : "PUBLIC")
                .scheduledAt(request.getScheduledAt())
                .isScheduled(true)
                .stats(new PostEntity.PostStats())
                .build();

        PostEntity savedPost = postRepository.save(post);
        return new PostResponse(savedPost);
    }

    /**
     * Enhanced search with filters
     */
    public List<PostResponse> searchPosts(String query, String category, String faculty,
                                        String dateRange, Pageable pageable) {
        // This would need to be implemented with proper MongoDB queries
        // For now, implementing basic search
        Page<PostEntity> posts;

        if (category != null && !category.isEmpty()) {
            posts = postRepository.findByCategoryAndTitleContainingOrContentContaining(
                category, query, query, pageable);
        } else {
            posts = postRepository.findByTitleContainingOrContentContaining(query, query, pageable);
        }

        return posts.stream()
                .map(PostResponse::new)
                .collect(Collectors.toList());
    }

    /**
     * Update author information in posts (for data consistency)
     */
    public void updateAuthorInfoInPosts(String userId, String fullName, String avatarUrl) {
        List<PostEntity> userPosts = postRepository.findByAuthor_Id(userId);

        for (PostEntity post : userPosts) {
            AuthorInfo updatedAuthor = post.getAuthor();
            updatedAuthor.setFullName(fullName);
            updatedAuthor.setAvatarUrl(avatarUrl);
            post.setAuthor(updatedAuthor);
        }

        if (!userPosts.isEmpty()) {
            postRepository.saveAll(userPosts);
        }
    }

    // Helper methods
    private void handleLikeInteraction(String postId, String userId, String reactionType,
                                     PostEntity post, AuthorInfo user) {
        InteractionEntity.ReactionType reaction =
            reactionType != null ?
            InteractionEntity.ReactionType.valueOf(reactionType.toUpperCase()) :
            InteractionEntity.ReactionType.LIKE;

        // Check if user already reacted
        Optional<InteractionEntity> existingInteraction = interactionRepository
                .findByPostIdAndAuthor_IdAndType(postId, userId, InteractionEntity.InteractionType.REACTION);

        if (existingInteraction.isEmpty()) {
            // Create new reaction
            InteractionEntity interaction = InteractionEntity.builder()
                    .postId(postId)
                    .author(user)
                    .type(InteractionEntity.InteractionType.REACTION)
                    .reactionType(reaction)
                    .build();

            interactionRepository.save(interaction);
            post.getStats().incrementReaction(reaction);
        }
    }

    private void handleUnlikeInteraction(String postId, String userId, PostEntity post) {
        Optional<InteractionEntity> existingInteraction = interactionRepository
                .findByPostIdAndAuthor_IdAndType(postId, userId, InteractionEntity.InteractionType.REACTION);

        if (existingInteraction.isPresent()) {
            InteractionEntity interaction = existingInteraction.get();
            post.getStats().decrementReaction(interaction.getReactionType());
            interactionRepository.delete(interaction);
        }
    }

    private void handleShareInteraction(String postId, String userId, PostEntity post, AuthorInfo user) {
        InteractionEntity interaction = InteractionEntity.builder()
                .postId(postId)
                .author(user)
                .type(InteractionEntity.InteractionType.SHARE)
                .build();

        interactionRepository.save(interaction);
        post.getStats().incrementShares();
    }

    private void updateEngagementMetrics(PostEntity post) {
        PostEntity.EngagementMetrics engagement = post.getEngagement();
        PostEntity.PostStats stats = post.getStats();

        engagement.updateEngagement(
            (int) stats.getLikes(),
            (int) stats.getComments(),
            (int) stats.getShares(),
            (int) stats.getViews()
        );
    }
}
